iam-token-manager/v1.ts

```diff
@@ -0,0 +1,198 @@
+        return cb(err, this.tokenInfo.access_token);
+      });
+    } else {
+      // 4. use valid managed token
+      return cb(null, this.tokenInfo.access_token);
+    }
+  }
+
+  /**
+   * Request an IAM token using an API key.
+   *
+   * @param {string} authHeader - The authorization header to send to the service. Defaults to 'Basic Yng6Yng='.
+   * @param {Function} cb - The callback that handles the response.
+   * @returns {void}
+   */
+  public requestToken(authHeader: string, cb: Function): void {
+    const authorization = authHeader || 'Basic Yng6Yng=';
+    const parameters = {
+      options: {
+        url: this.url,
+        method: 'POST',
+        headers: {
+          'Content-type': 'application/x-www-form-urlencoded',
+          Authorization: authorization
+        },
+        form: {
+          grant_type: 'urn:ibm:params:oauth:grant-type:apikey',
+          apikey: this.iam_apikey,
+          response_type: 'cloud_iam'
+        }
+      }
+    };
+    sendRequest(parameters, (error, body, response) => {
+      cb(error, body);
+    });
+  }
+
+  /**
+   * Refresh an IAM token using a refresh token.
+   *
+   * @param {string} authHeader - The authorization header to send to the service. Defaults to 'Basic Yng6Yng='.
+   * @param {Function} cb - The callback that handles the response.
+   * @returns {void}
+   */
+  public refreshToken(authHeader: string, cb: Function) {
+    const authorization = authHeader || 'Basic Yng6Yng=';
+    const parameters = {
+      options: {
+        url: this.url,
+        method: 'POST',
+        headers: {
+          'Content-type': 'application/x-www-form-urlencoded',
+          Authorization: authorization
+        },
+        form: {
+          grant_type: 'refresh_token',
+          refresh_token: this.tokenInfo.refresh_token
+        }
+      }
+    };
+    sendRequest(parameters, (error, body, response) => {
+      cb(error, body);
+    });
+  }
+
+  /**
+   * Set a self-managed IAM access token.
+   * The access token should be valid and not yet expired.
+   *
+   * By using this method, you accept responsibility for managing the
+   * access token yourself. You must set a new access token before this
+   * one expires. Failing to do so will result in authentication errors
+   * after this token expires.
+   *
+   * @param {string} iam_access_token - A valid, non-expired IAM access token
+   * @returns {void}
+   */
+  public setAccessToken(iam_access_token: string): void { // tslint:disable-line variable-name
+    this.userAccessToken = iam_access_token;
+  }
+
+  /**
+   * Check if currently stored token is expired.
+   * 
+   * Using a buffer to prevent the edge case of the 
+   * token expiring before the request could be made.
+   *
+   * The buffer will be a fraction of the total TTL. Using 80%.
+   *
+   * @private
+   * @returns {boolean}
+   */
+  private isTokenExpired(): boolean {
+    const fractionOfTtl = 0.8;
+    const timeToLive = this.tokenInfo.expires_in;
```

Comment:
What happens if `expires_in` is not defined?

Reply:
This function should never be called if `expires_in` is undefined. `expires_in` gets set at the same time as `access_token`, so the check for `!this.tokenInfo.accessToken` should trigger a request if they aren't defined.

Reply:
Might be straightforward to add a defensive check, just in case:

```node
if (!this.tokenInfo.expires_in || !this.tokenInfo.expiration) return true;
```